### PR TITLE
[Feature][Connector-V2] Support Druid Source & Sink

### PR DESCRIPTION
**Pull request by guanboo**
fix [Feature][Connector] New SeaTunnel API Connectors #1946 Add Druid Source & Sink

### CODE REVIEW

File: docs/en/connector-v2/sink/Druid.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# Druid
+
+> Druid sink connector
+
+## Description
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Druid.md:5]: Please add the `Key Features` reference https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/sink/Clickhouse.md

File: docs/en/connector-v2/sink/Druid.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+| name                    | type     | required | default value |
+| ----------------------- | -------- | -------- | ------------- |
+| coordinator_url         | `String` | yes      | -             |
+| datasource              | `String` | yes      | -             |
+| columns                 | `List<String>` | yes| __time        |
+| timestamp_column        | `String` | no       | timestamp     |
+| timestamp_format        | `String` | no       | auto          |
+| timestamp_missing_value | `String` | no       | -             |
+
+### coordinator_url [`String`]
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Druid.md:21]: Please use the same format as other documents  https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/sink/Clickhouse.md

File: docs/en/connector-v2/source/Druid.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Druid
+
+> Druid source connector
+
+## Description
```
Comment (EricJoy2048) [inline docs/en/connector-v2/source/Druid.md:5]: Same as the Sink document.

**Review by ashulin — COMMENTED** [review]
Expect the connector to complete.

And can you add e2e to seatunnel-connector-v2-e2e module? you can see https://github.com/apache/incubator-seatunnel/pull/2924.

File: pom.xml [hunk 4]
```diff
@@ -198,6 +198,12 @@
         <snappy-java.version>1.1.8.3</snappy-java.version>
         <checker.qual.version>3.10.0</checker.qual.version>
         <awaitility.version>4.2.0</awaitility.version>
+        <httpclient.version>4.5.13</httpclient.version>
+        <fastjson.version>1.2.83</fastjson.version>
```
Comment (EricJoy2048) [inline pom.xml:202]: Please manager the jars which used by connector in connector pom.xml like other connector.

File: pom.xml [hunk 5]
```diff
@@ -350,6 +356,43 @@
...
+
+            <dependency>
+                <groupId>org.apache.httpcomponents</groupId>
+                <artifactId>httpclient</artifactId>
+                <version>${httpclient.version}</version>
+            </dependency>
+
+            <dependency>
+                <groupId>com.alibaba</groupId>
+                <artifactId>fastjson</artifactId>
```
Comment (EricJoy2048) [inline pom.xml:368]: Suggest use jackson replace fastjson.

File: seatunnel-connectors-v2/connector-druid/pom.xml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            <artifactId>jackson-datatype-joda</artifactId>
+            <version>${jackson-datatype-joda.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.alibaba</groupId>
+            <artifactId>fastjson</artifactId>
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/pom.xml:35]: same as above.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    public ResultSetMetaData getResultSetMetaData() throws SQLException {
+        try {
+            quarySQL = getSQL();
+            connection = DriverManager.getConnection(druidSourceOptions.getUrl());
+            statement = connection.prepareStatement(quarySQL, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+            return statement.getMetaData();
+        } catch (SQLException se) {
+            throw new SQLException("ResultSetMetaData() failed." + se.getMessage(), se);
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:74]: Please use `ExceptionUtils.getMessage()` replace `se.getMessage()`

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    public void openInputFormat() {
+        try {
+            connection = DriverManager.getConnection(druidSourceOptions.getUrl());
+            statement = connection.prepareStatement(quarySQL, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+            resultSet = statement.executeQuery();
+            hasNext = resultSet.next();
+        } catch (SQLException se) {
+            throw new IllegalArgumentException("openInputFormat() failed." + se.getMessage(), se);
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:85]: Same as above.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        }
+    }
+
+    private String getSQL() throws SQLException {
+        String columns = COLUMNS_DEFAULT;
+        String startTimestamp = druidSourceOptions.getStartTimestamp();
+        String endTimestamp = druidSourceOptions.getEndTimestamp();
+        String dataSource = druidSourceOptions.getDatasource();
+        if (druidSourceOptions.getColumns() != null && druidSourceOptions.getColumns().size() > 0) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:94]: Suggest use `CollectionUtils.isEmpty(druidSourceOptions.getColumns())`

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    public SeaTunnelRow toInternal(ResultSet rs, SeaTunnelRowType rowTypeInfo) throws SQLException {
+        List<Object> fields = new ArrayList<>();
+        SeaTunnelDataType<?>[] seaTunnelDataTypes = rowTypeInfo.getFieldTypes();
+
+        for (int i = 1; i <= seaTunnelDataTypes.length; i++) {
+            Object seatunnelField;
+            SeaTunnelDataType<?> seaTunnelDataType = seaTunnelDataTypes[i - 1];
+            if (null == rs.getObject(i)) {
+                seatunnelField = null;
+            } else if (BasicType.BOOLEAN_TYPE.equals(seaTunnelDataType)) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:161]: use seaTunnelDataType.getSqlType() better.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+                seatunnelField = rs.getString(i);
+            } else if (LocalTimeType.LOCAL_DATE_TIME_TYPE.equals(seaTunnelDataType)) {
+                Timestamp ts = rs.getTimestamp(i, Calendar.getInstance(TimeZone.getTimeZone("UTC")));
+                LocalDateTime localDateTime = LocalDateTime.ofInstant(ts.toInstant(), ZoneId.of("UTC"));  // good
+                seatunnelField = localDateTime;
+            } else if (LocalTimeType.LOCAL_TIME_TYPE.equals(seaTunnelDataType)) {
+                seatunnelField = rs.getDate(i);
+            } else if (LocalTimeType.LOCAL_DATE_TYPE.equals(seaTunnelDataType)) {
+                seatunnelField = rs.getDate(i);
+            } else {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:185]: I see druid support `DECIMAL` from https://druid.apache.org/docs/latest/querying/sql-data-types.html.

Can you support `DECIMAL` here?

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+                throw new IllegalStateException("Unexpected value: " + seaTunnelDataType);
+            }
+
+            fields.add(seatunnelField);
+        }
+
+        return new SeaTunnelRow(fields.toArray());
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:192]: Each row will call this method, I suggest we minimize unnecessary memory overhead. So I think `Object[] fields = new Object[seaTunnelDataTypes.getTotalFields()]` is better.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    private SeaTunnelRowType initTableField() {
+        ArrayList<SeaTunnelDataType<?>> seaTunnelDataTypes = new ArrayList<>();
+        ArrayList<String> fieldNames = new ArrayList<>();
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:197]: Same as above.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidOutputFormat.java [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8))) {
+                StringBuilder response = new StringBuilder();
+                String responseLine;
+                while ((responseLine = br.readLine()) != null) {
+                    response.append(responseLine.trim());
+                }
+                LOGGER.info("Druid write task has been sent, and the response is {}", response.toString());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidOutputFormat.java:153]: Please don't use `e.printStackTrace()` and use logger to output exception messages.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSourceOptions.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    private String endTimestamp;
+    private List<String> columns;
+
+    private String partitionColumn;
+    private Long partitionUpperBound;
+    private Long partitionLowerBound;
+    private Integer parallelism;
+
+    public DruidSourceOptions(Config pluginConfig) {
+        this.url = pluginConfig.getString(DruidSourceConfig.URL);
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSourceOptions.java:44]: Please check whether params exists before get it.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSinkOptions.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    private String coordinatorURL;
+    private String datasource;
+    private String timestampColumn;
+    private String timestampFormat;
+    private String timestampMissingValue;
+    private List<String> columns;
+    private int parallelism;
+
+    public DruidSinkOptions(Config pluginConfig) {
+        this.coordinatorURL = pluginConfig.getString(DruidSinkConfig.COORDINATOR_URL);
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSinkOptions.java:46]: Please check whether params exists before get it.

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/source/DruidSource.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public void prepare(Config pluginConfig) throws PrepareFailException {
+        LOGGER.info("Druid source prepare");
+        try {
+            druidSourceOptions = new DruidSourceOptions(pluginConfig);
+            druidInputFormat = new DruidInputFormat(druidSourceOptions);
+            this.rowTypeInfo = druidInputFormat.getRowTypeInfo();
+        } catch (Exception e) {
+            throw new PrepareFailException("Druid", PluginType.SOURCE, e.toString());
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/source/DruidSource.java:61]: Replace `e.toString()` with `ExceptionUtils.getMessage(e)`

**Comment by guanboo** [comment]
> Expect the connector to complete.
> 
> And can you add e2e to seatunnel-connector-v2-e2e module? you can see #2924.

hi bro，development e2e testcase,There are some problems with my environment. this is message:
https://github.com/apache/incubator-seatunnel/issues/2876#issuecomment-1257417413

**Comment by EricJoy2048** [comment]
Hi, @guanboo  Did you fix the problems?

**Comment by guanboo** [comment]
> Hi, @guanboo Did you fix the problems?

Yes,done

**Comment by ic4y** [comment]
@guanboo please fix CI error and resolve conflicts

**Comment by laglangyue** [comment]
hi brother,did you continue this PR?
I am studying the reader connector of druid. The implementation method is somewhat different from yours. Through the raw json protocol. I think we can communicate

**Comment by MonsterChenzhuo** [comment]
@guanboo  Will this pr go ahead? If not, I would like to take over and finish it

**Comment by laglangyue** [comment]
> @guanboo  Will this pr go ahead? If not, I would like to take over and finish it

yes,brother. the pr is going,the connector is complete，but e2e is not completely, I has finished the reader of e2e,and guanboo try to finished the writer of e2e, but he is so busy to do it.I not similar with the writer for druid.

**Comment by laglangyue** [comment]
> @guanboo  Will this pr go ahead? If not, I would like to take over and finish it

if you so similar with druid you can continue it or another pr，guanboo and I are so busy for our work.

**Pull request closed without merging**